### PR TITLE
Add NavigationManager.OnLocationChanging (#14962)

### PR DESCRIPTION
**Pull request by mrpmorris**
- Added OnLocationChanging event to router
- Allow cancelling of navigation via args.PreventNavigation()

Addresses #14962

### CODE REVIEW

File: src/Components/Components/src/Routing/LocationChangingEventArgs.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    public class LocationChangingEventArgs : EventArgs
+    {
+        /// <summary>
+        /// Initializes a new instance of <see cref="LocationChangingEventArgs" />.
+        /// </summary>
+        /// <param name="location">The location of the navigation request.</param>
+        /// <param name="forceLoad">If true, the requested navigation will bypass client-side routing and force the browser to load the new page from the server, whether or not the URI would normally be handled by the client-side router.</param>
+        public LocationChangingEventArgs(string location, bool forceLoad)
+        {
+            Location = location;
```
Comment (campersau) [inline src/Components/Components/src/Routing/LocationChangingEventArgs.cs]:
```suggestion
            Location = location;
            ForceLoad = forceLoad;
```

Comment (mrpmorris) [reply src/Components/Components/src/Routing/LocationChangingEventArgs.cs]: Thank you!

**Comment by SteveSandersonMS** [comment]
Thanks for suggesting this, @mrpmorris.

Currently I don't quite follow what the intended design is here. This only runs `CanNavigateTo` when `NavigateToCore` is about to execute, which only happens when the developer triggers navigation programmatically using `NavigationManager.NavigateTo`. It will *not* have any effect on navigation triggered by clicking on `<a>` tags. As such it doesn't seem very useful - if a developer really wants to run special logic before programmatic navigation, they can just put that in their code directly before their call to `NavigationManager.NavigateTo`. If they really want to centralise this check in one place, they can create their own utility method that runs their own `CanNavigateTo` before calling `NavigationManager.NavigateTo`, so it doesn't need to be baked into `NavigationManager` itself.

I'm guessing it's intentional to only do this for programmatic navigation, since it wouldn't be possible to block navigation that occurs when people click on `<a>` tags. For Blazor Server, the logic has to run asynchronously so the JS event can't be cancelled. Event for Blazor WebAssembly or any JS-based SPA framework, it doesn't make sense to block clicks on `<a>` tags, since browsers won't always honour your intentions (e.g., if a user right-clicks and chooses *Open in new tab*).

**Comment by mrpmorris** [comment]
Sorry, I must have misunderstood. I thought the JS intercepted navigations and then pushed them through to Blazor.

If that's not the case then this PR is quite useless. I found it quite difficult to test the changes, so thanks for having a look!

**Pull request closed without merging**